core/ecschema-metadata/src/UnitQuery/UnitQuery.ts

```diff
@@ -0,0 +1,171 @@
+import { SchemaContext, SchemaItem, SchemaItemKey, SchemaKey, Unit } from "../ecschema-metadata";
+import { SchemaItemType } from "../ECObjects";
+import { UNIT_DATA } from "./UnitsData";
+
+export class UnitQuery {
+  /**
+   * @param _context
+   */
+  constructor(private readonly _context: SchemaContext) {}
+
+  /**
+   * Find unit in a schema that has unitName
+   * @param unitName Full name of unit
+   * @returns Unit whose full name matches unitName
+   */
+  public async findUnitByName(unitName: string): Promise<Unit> {
+    // Check if schema exists and unit exists in schema
+    const [schemaName, schemaItemName] = SchemaItem.parseFullName(unitName);
+    const schemaKey = new SchemaKey(schemaName);
+    const schema = await this._context.getSchema(schemaKey);
+
+    if (!schema) {
+      throw new BentleyError(BentleyStatus.ERROR, "Cannot find schema for unit", () => {
+        return { schema: schemaName, unit: unitName };
+      });
+    }
+
+    const itemKey = new SchemaItemKey(schemaItemName, schema.schemaKey);
+    const item = await this._context.getSchemaItem(itemKey);
+    if (!item)
+      throw new BentleyError(BentleyStatus.ERROR, "Cannot find schema item/unit", () => {
+        return { item: schemaItemName, schema: schemaName };
+      });
+
+    if (item.schemaItemType === SchemaItemType.Unit)
+      return item as Unit;
+
+    throw new BentleyError(BentleyStatus.ERROR, "Item is not a unit", () => {
+      return { itemType: item.key.fullName };
+    });
+  }
+
+  /**
+   * Find all units in context that belongs to given phenomenon
+   * @param phenomenon Full name of phenomenon
+   * @returns Array of units whose full name matches phenomenon param
+   */
+  public async findUnitsByPhenomenon(phenomenon: string): Promise<Array<Unit>> {
+    // Check if schema exists and phenomenon exists in schema
+    const [schemaName, schemaItemName] = SchemaItem.parseFullName(phenomenon);
+    const schemaKey = new SchemaKey(schemaName);
+    const schema = await this._context.getSchema(schemaKey);
+
+    if (!schema) {
+      throw new BentleyError(BentleyStatus.ERROR, "Cannot find schema for phenomenon", () => {
+        return { phenomenon, schema: schemaName };
+      });
+    }
+
+    const itemKey = new SchemaItemKey(schemaItemName, schema.schemaKey);
+    const phenom = await this._context.getSchemaItem(itemKey);
+    if (!phenom)
+      throw new BentleyError(BentleyStatus.ERROR, "Cannot find schema item/phenomenon", () => {
+        return { item: schemaItemName, schema: schemaName };
+      });
+
+    if (phenom.schemaItemType !== SchemaItemType.Phenomenon)
+      throw new BentleyError(BentleyStatus.ERROR, "Item is not a phenomenon", () => {
+        return { itemType: phenom.key.fullName };
+      });
+
+    // Filter out schema items to find units' full name that match given phenomenon param
+    const filteredUnits: Array<Unit> = [];
+    this._context.iterateSchemaItems((schemaItem) => {
```

Comment:
If you make the change referenced above about switching it to a generator this would be something like,

```ts
for (const item of this._context.items) {
  if (item.schemaItemType === SchemaItemType.Unit && schemaItem.phenomenon?.fullName === phenomenon)
    filteredUnits.push(schemaItem);
}
```

I think one other change that is needed to the matching here is to use `SchemaItem.Key.matchesFullName` rather than doing the fullName comparison yourself at the end of the line.  That'd look something like this,

```ts
schemaItem.phenomenon?.key.matchesFullName(phenomenon);
```

Reply:
Changed to use `SchemaItem.Key.matchesFullName`